File: backend/src/main/java/com/example/app/billing/InvoicePrintFormat.java
package com.example.app.billing;

import java.util.Locale;

/** Output format used when generating an invoice document for printing. */
public enum InvoicePrintFormat {
    A4,
    POS_58;

    public static InvoicePrintFormat from(String raw) {
        if (raw == null || raw.isBlank()) return A4;
        String normalized = raw.trim().toUpperCase(Locale.ROOT)
                .replace('-', '_')
                .replace(' ', '_');
        if (normalized.equals("POS58") || normalized.equals("58MM") || normalized.equals("POS_58MM")) {
            return POS_58;
        }
        try {
            return valueOf(normalized);
        } catch (IllegalArgumentException ignored) {
            return A4;
        }
    }
}

File: frontend/src/lib/invoicePrintFormat.ts
export const DEFAULT_INVOICE_PRINT_FORMAT_KEY = 'DEFAULT_INVOICE_PRINT_FORMAT'

export type InvoicePrintFormat = 'A4' | 'POS_58'
export type InvoicePrintPreference = InvoicePrintFormat | 'ASK'

export function normalizeInvoicePrintPreference(value: unknown): InvoicePrintPreference {
  const normalized = String(value ?? '').trim().toUpperCase().replace(/[\s-]+/g, '_')
  if (normalized === 'POS58' || normalized === 'POS_58MM' || normalized === '58MM') return 'POS_58'
  if (normalized === 'POS_58' || normalized === 'ASK') return normalized
  return 'A4'
}